pegen/pegen.c

```diff
@@ -503,3 +503,64 @@ seq_flatten(Parser *p, asdl_seq *seqs)
 
     return flattened_seq;
 }
+
+expr_ty
+seq_to_dotted_name(Parser *p, asdl_seq *seq)
+{
+    /* Create a string of the form a.b.c like Python/ast.c alias_for_dotted_name does
+       for dotted names */
+    if (asdl_seq_LEN(seq) == 1) {
+        return asdl_seq_GET(seq, 0);
+    }
+
+    // Get length of dotted string
+    ssize_t len = 0;
+    for (int i = 0, l = asdl_seq_LEN(seq); i < l; i++) {
+        expr_ty current_expr = asdl_seq_GET(seq, i);
+        ssize_t current_len = PyUnicode_GET_LENGTH(current_expr->v.Name.id);
+        len += current_len + 1;
+    }
+    --len; // Last name does not have a dot
+
+    PyObject *str = PyBytes_FromStringAndSize(NULL, len);
+    if (!str) {
+        return NULL;
+    }
+
+    char *s = PyBytes_AS_STRING(str);
+    if (!s) {
+        return NULL;
+    }
+
+    for (int i = 0, l = asdl_seq_LEN(seq); i < l; i++) {
+        expr_ty current_expr = asdl_seq_GET(seq, i);
+        const char *current_name = PyUnicode_AsUTF8(current_expr->v.Name.id);
+        if (!current_name) {
+            return NULL;
+        }
+        strcpy(s, current_name);
+        s += strlen(current_name);
+        *s++ = '.';
+    }
+    s--;
+    *s = '\0';
+
+    PyObject *uni = PyUnicode_DecodeUTF8(PyBytes_AS_STRING(str),
+                                         PyBytes_GET_SIZE(str),
+                                         NULL);
+    Py_DECREF(str);
+    if (!uni) {
+        return NULL;
+    }
+    str = uni;
+    PyUnicode_InternInPlace(&str);
+    if (PyArena_AddPyObject(p->arena, str) < 0) {
+        Py_DECREF(str);
+        return NULL;
+    }
+
+    expr_ty first_elem = asdl_seq_GET(seq, 0);
+    expr_ty last_elem = asdl_seq_GET(seq, asdl_seq_LEN(seq) - 1);
+    return Name(str, Load, first_elem->lineno, first_elem->col_offset,
+                last_elem->end_lineno, last_elem->end_col_offset, p->arena);
+}
```

Comment:
All in all this function looks very complex and error-prone. I wonder if we can't do better? One possibility would be to have just a function that takes two NAME tokens and joins them with a dot, and rewrite the grammar to be recursive instead of iterative. If there are a lot of dots that's more expensive, but I doubt that this would cost a measurable amount.

Reply:
I agree that this version seems error-prone and that a recursive rule might work a bit better here. I was thinking about that, while I was writing it.

I am not sure if this is a valid concern, but in case we publish the grammar on the Docs like with the current one and we want to make it easier for people to understand it, making a recursive grammar for this might not be the best idea. In that case, I think that a better version of this function might just be good enough.

Reply:
I honestly think that right now we ought to concentrate on getting a valid AST for all Python programs. We'll rethink the display value of the grammar later. If you want to you create an issue to remind us to consider that separately.

Reply:
Do you think I should open a new PR for this new recursive grammar or just commit to this branch?

Reply:
I find the current `seq_to_dotted_name()` implementation too fragile, so I'd rather make that change now.